# 项目：评估和清理英国电商公司销售数据

## 分析目标

此数据分析的目的是，根据市场销售数据，挖掘畅销产品，以便制定更有效的市场策略来提升营收。

本实战项目的目的在于练习评估数据干净和整洁度，并且基于评估结果，对数据进行清洗，从而得到可供下一步分析的数据。

## 简介

原始数据集记录了一家英国在线零售公司在2010年12月1日至2011年12月9日期间的所有交易情况，涵盖了该公司在全球不同国家和地区的业务数据。该公司主要销售覆盖各个场景的礼品，包括但不限于生日礼品、结婚纪念品、圣诞礼品等等。该公司的客户群体主要包括批发商和个人消费者，其中批发商占据了相当大的比例。

数据每列的含义如下：
- `InvoiceNo`: 发票号码。6位数，作为交易的唯一标识符。如果这个代码以字母“c”开头，表示这笔交易被取消。
- `StockCode`: 产品代码。5位数，作为产品的唯一标识符。
- `Description`: 产品名称。
- `Quantity`: 产品在交易中的数量。
- `InvoiceDate`: 发票日期和时间。交易发生的日期和时间。
- `UnitPrice`: 单价。价格单位为英镑（£）。
- `CustomerID`: 客户编号。5位数，作为客户的唯一标识符。
- `Country`: 国家名称。客户所居住的国家的名称。

## 读取数据

导入`pandas模块`取别名为`as`，使用`.read_scv()方法`读取`csv格式文件`并保存在`orighinal_data`变量中。
使用`.sample(5)`随机打印了5行读取。

In [9]:
import pandas as pd

In [10]:
orighinal_data = pd.read_csv("./e_commerce.csv")

In [11]:
orighinal_data.sample(5)

InvoiceNo StockCode                          Description  Quantity  \
5497      536863     21452                  TOADSTOOL MONEY BOX         1   
460801    575930     22028         PENNY FARTHING BIRTHDAY CARD         1   
327147    565617     85048  15CM CHRISTMAS GLASS BALL 20 LIGHTS         1   
166938    550957     20719               WOODLAND CHARLOTTE BAG         1   
479126    577141     22752         SET 7 BABUSHKA NESTING BOXES         4   

             InvoiceDate  UnitPrice  CustomerID         Country  
5497     12/3/2010 11:19       2.95     17967.0  United Kingdom  
460801  11/11/2011 17:58       0.83         NaN  United Kingdom  
327147    9/5/2011 15:30      16.63         NaN  United Kingdom  
166938   4/21/2011 16:37       0.85     18283.0  United Kingdom  
479126   11/18/2011 9:11       8.50     13269.0  United Kingdom

## 评估数据

In [13]:
# 评估数据整洁度 head() tail() sample()
orighinal_data.tail()

InvoiceNo StockCode                      Description  Quantity  \
541904    581587     22613      PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899     CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254    CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255  CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138    BAKING SET 9 PIECE RETROSPOT          3   

            InvoiceDate  UnitPrice  CustomerID Country  
541904  12/9/2011 12:50       0.85     12680.0  France  
541905  12/9/2011 12:50       2.10     12680.0  France  
541906  12/9/2011 12:50       4.15     12680.0  France  
541907  12/9/2011 12:50       4.15     12680.0  France  
541908  12/9/2011 12:50       4.95     12680.0  France

从抽样的10行数据来看，数据符合"每列是一个变量，每行是一个观察值，每个单元格是一个值"。
具体来看每行是关于某电商品的一次交易，每列是交易相关的各个变量，因此不存在结构性问题。

In [15]:
# 评估数据干净程度 info()
orighinal_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


从`info()`方法输出结果来看，数据共有541909条观察值,而`Description CustomerID`存在缺失值。
此外`InvoiceDate`的数据类型应该是日期，`CustomerID`的数据类型应为字符串，应当进行数据格式转换。

#### 评估缺失值数据

In [18]:
# .isnull()查看数据的缺失值,为 NaN返回 False
# 查看"Description"的缺失值
orighinal_data[orighinal_data["Description"].isnull()]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
622       536414     22139         NaN        56  12/1/2010 11:52        0.0   
1970      536545     21134         NaN         1  12/1/2010 14:32        0.0   
1971      536546     22145         NaN         1  12/1/2010 14:33        0.0   
1972      536547     37509         NaN         1  12/1/2010 14:33        0.0   
1987      536549    85226A         NaN         1  12/1/2010 14:34        0.0   
...          ...       ...         ...       ...              ...        ...   
535322    581199     84581         NaN        -2  12/7/2011 18:26        0.0   
535326    581203     23406         NaN        15  12/7/2011 18:31        0.0   
535332    581209     21620         NaN         6  12/7/2011 18:35        0.0   
536981    581234     72817         NaN        27  12/8/2011 10:33        0.0   
538554    581408     85175         NaN        20  12/8/2011 14:06        0.0   

        CustomerID         Country  
622            NaN  United Kingdom  
1970           NaN  United Kingdom  
1971           NaN  United Kingdom  
1972           NaN  United Kingdom  
1987           NaN  United Kingdom  
...            ...             ...  
535322         NaN  United Kingdom  
535326         NaN  United Kingdom  
535332         NaN  United Kingdom  
536981         NaN  United Kingdom  
538554         NaN  United Kingdom  

[1454 rows x 8 columns]

In [19]:
# 继续查看是否有 "Description"为`NaN`的同时，`UnitPrice`不为零的情况
orighinal_data[(orighinal_data["Description"].isnull()) & (orighinal_data["UnitPrice"] != 0.0)]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

更具分析目标来判断  `Description: 产品名称。`和 `UnitPrice: 单价。价格单位为英镑（£）。`为`NaN`的数据不具备判断条件，应该作为垃圾数据清洗。

In [21]:
# 查看`CustomerID`的缺失值
orighinal_data[orighinal_data["CustomerID"].isnull()]

InvoiceNo StockCode                      Description  Quantity  \
622       536414     22139                              NaN        56   
1443      536544     21773  DECORATIVE ROSE BATHROOM BOTTLE         1   
1444      536544     21774  DECORATIVE CATS BATHROOM BOTTLE         2   
1445      536544     21786               POLKADOT RAIN HAT          4   
1446      536544     21787            RAIN PONCHO RETROSPOT         2   
...          ...       ...                              ...       ...   
541536    581498    85099B          JUMBO BAG RED RETROSPOT         5   
541537    581498    85099C   JUMBO  BAG BAROQUE BLACK WHITE         4   
541538    581498     85150    LADIES & GENTLEMEN METAL SIGN         1   
541539    581498     85174                S/4 CACTI CANDLES         1   
541540    581498       DOT                   DOTCOM POSTAGE         1   

            InvoiceDate  UnitPrice  CustomerID         Country  
622     12/1/2010 11:52       0.00         NaN  United Kingdom  
1443    12/1/2010 14:32       2.51         NaN  United Kingdom  
1444    12/1/2010 14:32       2.51         NaN  United Kingdom  
1445    12/1/2010 14:32       0.85         NaN  United Kingdom  
1446    12/1/2010 14:32       1.66         NaN  United Kingdom  
...                 ...        ...         ...             ...  
541536  12/9/2011 10:26       4.13         NaN  United Kingdom  
541537  12/9/2011 10:26       4.13         NaN  United Kingdom  
541538  12/9/2011 10:26       4.96         NaN  United Kingdom  
541539  12/9/2011 10:26      10.79         NaN  United Kingdom  
541540  12/9/2011 10:26    1714.17         NaN  United Kingdom  

[135080 rows x 8 columns]

In [22]:
# 继续查看有同时 "CustomerID"为空 ,"UnitPrice"为空,"Description"为空 的情况。
orighinal_data[orighinal_data["CustomerID"].isnull() & orighinal_data["CustomerID"].isnull() & orighinal_data["Description"].isnull()]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
622       536414     22139         NaN        56  12/1/2010 11:52        0.0   
1970      536545     21134         NaN         1  12/1/2010 14:32        0.0   
1971      536546     22145         NaN         1  12/1/2010 14:33        0.0   
1972      536547     37509         NaN         1  12/1/2010 14:33        0.0   
1987      536549    85226A         NaN         1  12/1/2010 14:34        0.0   
...          ...       ...         ...       ...              ...        ...   
535322    581199     84581         NaN        -2  12/7/2011 18:26        0.0   
535326    581203     23406         NaN        15  12/7/2011 18:31        0.0   
535332    581209     21620         NaN         6  12/7/2011 18:35        0.0   
536981    581234     72817         NaN        27  12/8/2011 10:33        0.0   
538554    581408     85175         NaN        20  12/8/2011 14:06        0.0   

        CustomerID         Country  
622            NaN  United Kingdom  
1970           NaN  United Kingdom  
1971           NaN  United Kingdom  
1972           NaN  United Kingdom  
1987           NaN  United Kingdom  
...            ...             ...  
535322         NaN  United Kingdom  
535326         NaN  United Kingdom  
535332         NaN  United Kingdom  
536981         NaN  United Kingdom  
538554         NaN  United Kingdom  

[1454 rows x 8 columns]

`CustomerID`: 客户编号。5位数，作为客户的唯一标识符。
`UnitPrice`: 单价。价格单位为英镑（£）。
`Description`: 产品名称。
根据分析 ，上面这三个参数同时为空,不会对最终`畅销产品分析`有实际效果，应当做垃圾数据进行清洗。

#### 评估重复数据
- `InvoiceNo`: 发票号码。6位数，作为交易的唯一标识符。如果这个代码以字母“c”开头，表示这笔交易被取消。
- `StockCode`: 产品代码。5位数，作为产品的唯一标识符。
- `CustomerID`: 客户编号。5位数，作为客户的唯一标识符。

In [25]:
# 同时查看 InvoiceNo StockCode CustomerID 这些唯一表示是否重复 (重复为True)
orighinal_data.duplicated(subset=["InvoiceNo","StockCode","CustomerID"])
# 打印这些重复数据
orighinal_data[orighinal_data.duplicated(subset=["InvoiceNo","StockCode","CustomerID"])]

InvoiceNo StockCode                      Description  Quantity  \
125       536381     71270                  PHOTO CLIP LINE         3   
498       536409    90199C  5 STRAND GLASS NECKLACE CRYSTAL         1   
502       536409     85116  BLACK CANDELABRA T-LIGHT HOLDER         5   
517       536409     21866      UNION JACK FLAG LUGGAGE TAG         1   
525       536409    90199C  5 STRAND GLASS NECKLACE CRYSTAL         2   
...          ...       ...                              ...       ...   
541692    581538     22992           REVOLVER WOODEN RULER          1   
541697    581538     21194        PINK  HONEYCOMB PAPER FAN         1   
541698    581538    35004B      SET OF 3 BLACK FLYING DUCKS         1   
541699    581538     22694                     WICKER STAR          1   
541701    581538     23343     JUMBO BAG VINTAGE CHRISTMAS          1   

            InvoiceDate  UnitPrice  CustomerID         Country  
125      12/1/2010 9:41       1.25     15311.0  United Kingdom  
498     12/1/2010 11:45       6.35     17908.0  United Kingdom  
502     12/1/2010 11:45       2.10     17908.0  United Kingdom  
517     12/1/2010 11:45       1.25     17908.0  United Kingdom  
525     12/1/2010 11:45       6.35     17908.0  United Kingdom  
...                 ...        ...         ...             ...  
541692  12/9/2011 11:34       1.95     14446.0  United Kingdom  
541697  12/9/2011 11:34       0.65     14446.0  United Kingdom  
541698  12/9/2011 11:34       5.45     14446.0  United Kingdom  
541699  12/9/2011 11:34       2.10     14446.0  United Kingdom  
541701  12/9/2011 11:34       2.08     14446.0  United Kingdom  

[10684 rows x 8 columns]

根据数据变量的含义来看,虽然`InvolceNo`,`StockCode`,`CustomerID`都是唯一标识。但是一次交易可能包含多见商品，因此`InvoiceNo`: 发票号码可以存在重复；不同交易可以包含同件商品，因此`StockCode`: 产品代码可以重复存在；顾客可以进行多次交易或下单多个商品,因此`CustomerID`: 客户编号也可以重复存在。
那么正对此数据集，我们无需评估重复数据。

#### 评估不一致数据
可以看看是否有不同字符串取代同一个国家的情况

In [55]:
# Country: 国家名称。客户所居住的国家的名称。
orighinal_data["Country"].value_counts()

Country
United Kingdom          495266
Germany                   9495
France                    8557
EIRE                      8196
Spain                     2533
Netherlands               2371
Belgium                   2069
Switzerland               2002
Portugal                  1519
Australia                 1259
Norway                    1086
Italy                      803
Channel Islands            758
Finland                    695
Cyprus                     622
Sweden                     462
Unspecified                446
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     297
China                      288
Singapore                  229
USA                        218
UK                         211
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United States               73
United Arab Emirates        68


从`Country`变量值来看,`USA`,`United States`均在表示美国，`United Kingdom`,`UK`,`U.K.`均在表示英国。因此应该在清洗步骤对这些值进行统一。只保留一个取代值。

#### 评估无效或错误的数据
可以通过`DataFrame对象`的`describe`方法，对数值统计进行快速了解。

In [59]:
orighinal_data.describe()

Quantity      UnitPrice     CustomerID
count  541909.000000  541909.000000  406829.000000
mean        9.552250       4.611114   15287.690570
std       218.081158      96.759853    1713.600303
min    -80995.000000  -11062.060000   12346.000000
25%         1.000000       1.250000   13953.000000
50%         3.000000       2.080000   15152.000000
75%        10.000000       4.130000   16791.000000
max     80995.000000   38970.000000   18287.000000

可以发现 `Quantity: 产品在交易中的数量` 和 `UnitPrice: 单价。价格单位为英镑（£）`存在负数。按照原则来看，数量和单价不能出现负数情况。
可以先筛选出为负数的数据看看。

In [73]:
# 查看交易数量为负数(小于0)情况
orighinal_data[orighinal_data["Quantity"] <= 0]

InvoiceNo StockCode                       Description  Quantity  \
141      C536379         D                          Discount        -1   
154      C536383    35004C   SET OF 3 COLOURED  FLYING DUCKS        -1   
235      C536391     22556    PLASTERS IN TIN CIRCUS PARADE        -12   
236      C536391     21984  PACK OF 12 PINK PAISLEY TISSUES        -24   
237      C536391     21983  PACK OF 12 BLUE PAISLEY TISSUES        -24   
...          ...       ...                               ...       ...   
540449   C581490     23144   ZINC T-LIGHT HOLDER STARS SMALL       -11   
541541   C581499         M                            Manual        -1   
541715   C581568     21258        VICTORIAN SEWING BOX LARGE        -5   
541716   C581569     84978  HANGING HEART JAR T-LIGHT HOLDER        -1   
541717   C581569     20979     36 PENCILS TUBE RED RETROSPOT        -5   

            InvoiceDate  UnitPrice  CustomerID         Country  
141      12/1/2010 9:41      27.50     14527.0  United Kingdom  
154      12/1/2010 9:49       4.65     15311.0  United Kingdom  
235     12/1/2010 10:24       1.65     17548.0  United Kingdom  
236     12/1/2010 10:24       0.29     17548.0  United Kingdom  
237     12/1/2010 10:24       0.29     17548.0  United Kingdom  
...                 ...        ...         ...             ...  
540449   12/9/2011 9:57       0.83     14397.0  United Kingdom  
541541  12/9/2011 10:28     224.69     15498.0  United Kingdom  
541715  12/9/2011 11:57      10.95     15311.0  United Kingdom  
541716  12/9/2011 11:58       1.25     17315.0  United Kingdom  
541717  12/9/2011 11:58       1.25     17315.0  United Kingdom  

[10624 rows x 8 columns]

In [93]:
# 查看 `Quantity: 产品在交易中的数量`为负数 并且 `InvoiceNo: 发票号码。6位数，作为交易的唯一标识符。如果这个代码以字母“c”开头，表示这笔交易被取消。`的数据
orighinal_data[(orighinal_data["Quantity"]<0) & (orighinal_data["InvoiceNo"].str[0] != "C")]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
2406      536589     21777         NaN       -10  12/1/2010 16:50        0.0   
4347      536764    84952C         NaN       -38  12/2/2010 14:42        0.0   
7188      536996     22712         NaN       -20  12/3/2010 15:30        0.0   
7189      536997     22028         NaN       -20  12/3/2010 15:30        0.0   
7190      536998     85067         NaN        -6  12/3/2010 15:30        0.0   
...          ...       ...         ...       ...              ...        ...   
535333    581210     23395       check       -26  12/7/2011 18:36        0.0   
535335    581212     22578        lost     -1050  12/7/2011 18:38        0.0   
535336    581213     22576       check       -30  12/7/2011 18:38        0.0   
536908    581226     23090     missing      -338   12/8/2011 9:56        0.0   
538919    581422     23169     smashed      -235  12/8/2011 15:24        0.0   

        CustomerID         Country  
2406           NaN  United Kingdom  
4347           NaN  United Kingdom  
7188           NaN  United Kingdom  
7189           NaN  United Kingdom  
7190           NaN  United Kingdom  
...            ...             ...  
535333         NaN  United Kingdom  
535335         NaN  United Kingdom  
535336         NaN  United Kingdom  
536908         NaN  United Kingdom  
538919         NaN  United Kingdom  

[1336 rows x 8 columns]

In [99]:
# ["Quantity"]<0 并且 ["InvoiceNo"].str[0] != "C" 并且 ["UnitPrice"] == 0.0
orighinal_data[(orighinal_data["Quantity"]<0) & (orighinal_data["InvoiceNo"].str[0] != "C") & (orighinal_data["UnitPrice"] == 0.0)]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
2406      536589     21777         NaN       -10  12/1/2010 16:50        0.0   
4347      536764    84952C         NaN       -38  12/2/2010 14:42        0.0   
7188      536996     22712         NaN       -20  12/3/2010 15:30        0.0   
7189      536997     22028         NaN       -20  12/3/2010 15:30        0.0   
7190      536998     85067         NaN        -6  12/3/2010 15:30        0.0   
...          ...       ...         ...       ...              ...        ...   
535333    581210     23395       check       -26  12/7/2011 18:36        0.0   
535335    581212     22578        lost     -1050  12/7/2011 18:38        0.0   
535336    581213     22576       check       -30  12/7/2011 18:38        0.0   
536908    581226     23090     missing      -338   12/8/2011 9:56        0.0   
538919    581422     23169     smashed      -235  12/8/2011 15:24        0.0   

        CustomerID         Country  
2406           NaN  United Kingdom  
4347           NaN  United Kingdom  
7188           NaN  United Kingdom  
7189           NaN  United Kingdom  
7190           NaN  United Kingdom  
...            ...             ...  
535333         NaN  United Kingdom  
535335         NaN  United Kingdom  
535336         NaN  United Kingdom  
536908         NaN  United Kingdom  
538919         NaN  United Kingdom  

[1336 rows x 8 columns]

`["Quantity"]<0` 并且 `["InvoiceNo"].str[0] != "C"` 并且 `["UnitPrice"] == 0.0` 的数据都是不贡献销售的，而且还会影响数据分析。在数据清理步骤中将其删除

In [77]:
# 查看单价小于0的情况
orighinal_data[orighinal_data["UnitPrice"] < 0]

InvoiceNo StockCode      Description  Quantity      InvoiceDate  \
299983   A563186         B  Adjust bad debt         1  8/12/2011 14:51   
299984   A563187         B  Adjust bad debt         1  8/12/2011 14:52   

        UnitPrice  CustomerID         Country  
299983  -11062.06         NaN  United Kingdom  
299984  -11062.06         NaN  United Kingdom

冲输出结果来看， `UnitPrice`为负数的观察值都是`坏账调整 Adjust bad debt`不属于实际商品交易数据,因此也在数据清理步骤中将其删除

## 清理数据

根据前面评估部分得到的结论,我们需要进行的数据清理包括:
1.（`InvoiceDate`: 发票日期和时间） 变量数据类型转换为日期时间
2.（`CustomerID`: 客户编号） 转换为字符串 
3.（`Description`: 产品名称）变量缺失的观察值删除
4.（`Country`: 国家名称） `USA`变成-->`United States` `UK`,`U.K.`变成-->`United Kingdom`
5.（`["Quantity"]<0 并且 ["InvoiceNo"].str[0] != "C" 并且 ["UnitPrice"] == 0.0`） 删除
6.（``UnitPrice`: 单价`）小于0的删除

In [271]:
# 为了区分开经过清理后数据和原始数据.我们创建新的变量`cleaned_data`,让它为`original_data`复制出的副本。
# 我们之后的清理步骤都将被运用在`cleaned_data`上.
cleand_data = orighinal_data.loc[:,:]

In [273]:
# 1.（`InvoiceDate`: 发票日期和时间） 变量数据类型转换为日期时间
# 使用 `pandas.to_datatime("需要转换日期时间的列")` 函数转换。此函数不会产生新数据。
cleand_data["InvoiceDate"] = pd.to_datetime(cleand_data["InvoiceDate"])
cleand_data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice  CustomerID         Country  
0      2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2      2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
...                    ...        ...         ...             ...  
541904 2011-12-09 12:50:00       0.85     12680.0          France  
541905 2011-12-09 12:50:00       2.10     12680.0          France  
541906 2011-12-09 12:50:00       4.15     12680.0          France  
541907 2011-12-09 12:50:00       4.15     12680.0          France  
541908 2011-12-09 12:50:00       4.95     12680.0          France  

[541909 rows x 8 columns]

In [275]:
# 2.（`CustomerID`: 客户编号） 转换为字符串
cleand_data["CustomerID"] = cleand_data["CustomerID"].astype(str)

# 把字符串`17850.0` 的 `.0`删除
# 字符串 str属性下有个 slice方法,可以让只保留字符串的一部分
cleand_data["CustomerID"] = cleand_data["CustomerID"].str.slice(0,-2)   # .str使用字符串属性(这样就能获得很多字符串的方法) .slice()方法参数 == range()参数的逻辑

In [277]:
cleand_data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice CustomerID         Country  
0      2010-12-01 08:26:00       2.55      17850  United Kingdom  
1      2010-12-01 08:26:00       3.39      17850  United Kingdom  
2      2010-12-01 08:26:00       2.75      17850  United Kingdom  
3      2010-12-01 08:26:00       3.39      17850  United Kingdom  
4      2010-12-01 08:26:00       3.39      17850  United Kingdom  
...                    ...        ...        ...             ...  
541904 2011-12-09 12:50:00       0.85      12680          France  
541905 2011-12-09 12:50:00       2.10      12680          France  
541906 2011-12-09 12:50:00       4.15      12680          France  
541907 2011-12-09 12:50:00       4.15      12680          France  
541908 2011-12-09 12:50:00       4.95      12680          France  

[541909 rows x 8 columns]

In [279]:
# 3.（`Description`: 产品名称）变量缺失的观察值删除
cleand_data = cleand_data.dropna(subset=["Description"])

In [343]:
# 打印不为空
cleand_data[cleand_data["Description"].notnull()]
# 查看"Description"为空值的和
cleand_data["Description"].isnull().sum()

0

In [345]:
# 4.（`Country`: 国家名称） `USA`变成-->`United States` `UK`,`U.K.`变成-->`United Kingdom`

#  `USA`变成-->`United States`
#一: 先定位所有 `USA`值
#二: 用.loc[] 方法定位从行开始 定位"Country"列下所有"USA"元素值 并 替换成" United States"
cleand_data.loc[cleand_data["Country"] == "USA","Country"] = "United States"
#三: 查看
cleand_data[cleand_data["Country"] == "United States"]
len(cleand_data[cleand_data["Country"] == "USA"])


# 修改 `UK`,`U.K.`变成-->`United Kingdom`
cleand_data["Country"].replace({"UK":"United Kingdom","U.K.":"United Kingdom"})
cleand_data.loc[(cleand_data["Country"] == "UK") | (cleand_data["Country"] == "U.K."),"Country"] = "United Kingdom"
# 查看
len(cleand_data[(cleand_data["Country"] == "UK") | (cleand_data["Country"] == "U.K.")])

In [351]:
# 重新打印 "Country" 列 的每行数据
cleand_data["Country"].value_counts()

Country
United Kingdom          494024
Germany                   9495
France                    8557
EIRE                      8196
Spain                     2533
Netherlands               2371
Belgium                   2069
Switzerland               2002
Portugal                  1519
Australia                 1259
Norway                    1086
Italy                      803
Channel Islands            758
Finland                    695
Cyprus                     622
Sweden                     462
Unspecified                446
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     297
United States              291
China                      288
Singapore                  229
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United Arab Emirates        68
European Community          61
RSA                         58


In [404]:
# 5.（`["Quantity"]<0 并且 ["InvoiceNo"].str[0] != "C" 并且 ["UnitPrice"] == 0.0`） 删除
_ = cleand_data[(cleand_data["Quantity"] < 0) & (cleand_data["InvoiceNo"].str[0] == "C") & (cleand_data["UnitPrice"] == 0.0)].index
cleand_data = cleand_data.drop(_)


In [406]:
# 查看
len(cleand_data[(cleand_data["Quantity"] < 0) & (cleand_data["InvoiceNo"].str[0] == "C") & (cleand_data["UnitPrice"] == 0.0)])

0

In [424]:
# 6.（`UnitPrice`: 单价`）小于0的删除  `.index`保存行索引
_ = cleand_data[cleand_data["UnitPrice"] < 0].index
# 把删除赋值
cleand_data = cleand_data.drop(_)

In [426]:
# 查看
len(cleand_data[cleand_data["UnitPrice"] < 0])

0

###### 其实删除某些东西还有一种思路，比如把满足某些条件的数据删除掉(我们可以把不满足这些删除条件的数据保存并使用)，就是变相的删除了不满足条件的数据。
在赛选数据前面加上 取反符号即可 ~
`~((cleand_data["Quantity"] < 0) & (cleand_data["InvoiceNo"].str[0] == "C") & (cleand_data["UnitPrice"] == 0.0))`

## 保存清理后的数据

保存数据命名: 无特殊要求，默认保存数据是在原始数据后面加上`_cleaned`用于和原始数据做出区分。

In [416]:
# 保存数据，并不保存位置索引
cleand_data.to_csv("./e_commerce_cleaned.csv",index = False)

In [422]:
# 打开保存数据 并且查看前5行
pd.read_csv("./e_commerce_cleaned.csv").head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

           InvoiceDate  UnitPrice CustomerID         Country  
0  2010-12-01 08:26:00       2.55      17850  United Kingdom  
1  2010-12-01 08:26:00       3.39      17850  United Kingdom  
2  2010-12-01 08:26:00       2.75      17850  United Kingdom  
3  2010-12-01 08:26:00       3.39      17850  United Kingdom  
4  2010-12-01 08:26:00       3.39      17850  United Kingdom